cpp/src/lists/contains.cu

```diff
@@ -46,10 +52,52 @@ auto get_search_keys_device_iterable_view(cudf::scalar const& search_key, rmm::c
   return &search_key;
 }
 
+/**
+ * @brief Enum to indicate whether the `search_key` scalar/column contains nulls.
+ */
+enum search_key_nulls : bool { NO_NULLS = false, HAS_NULLS = true };
+
+/**
+ * @brief __device__ functor to search for a key in a `list_device_view`.
+ */
+template <duplicate_find_option = duplicate_find_option::FIND_FIRST>
+struct finder {
+  template <typename ElementType>
+  __device__ size_type operator()(list_device_view const& list, ElementType const& search_key) const
+  {
+    auto const list_begin = list.pair_rep_begin<ElementType>();
+    auto const list_end   = list.pair_rep_end<ElementType>();
+    auto const find_iter  = thrust::find_if(
+      thrust::seq, list_begin, list_end, [search_key] __device__(auto element_and_validity) {
+        return element_and_validity.second &&
+               cudf::equality_compare(element_and_validity.first, search_key);
+      });
+    auto const is_found = find_iter != list_end;
+    return is_found ? (find_iter - list_begin) : absent_index;
+  };
+};
+
+template <>
+struct finder<duplicate_find_option::FIND_LAST> {
+  template <typename ElementType>
+  __device__ size_type operator()(list_device_view const& list, ElementType const& search_key) const
+  {
+    auto const begin = thrust::make_reverse_iterator(list.pair_rep_end<ElementType>());
+    auto const end   = thrust::make_reverse_iterator(list.pair_rep_begin<ElementType>());
+    auto const find_iter =
+      thrust::find_if(thrust::seq, begin, end, [search_key] __device__(auto element_and_validity) {
+        return element_and_validity.second &&
+               cudf::equality_compare(element_and_validity.first, search_key);
+      });
+    auto const is_found = find_iter != end;
+    return is_found ? (end - find_iter - 1) : absent_index;
+  };
+};
```

Comment:
Feels like these two could be merged and save some duplicated code...  Are list_begin/begin different types?

Reply:
They are different here, but we could make the iterator type a template parameter. 

I can take another pass at this. It does seem possible to reduce this, but I couldn't the last time I tried.

Reply:
It took a while to sort this one out.